File: programmers/level_2/오픈채팅방.py
def solution(record):
    answer = []
    tmp_answer = []
    mapping = {}
    for rec in record:
        rec = rec.split(' ')
        if len(rec) == 2:  # leave
            tmp_answer.append([rec[1], rec[0]])
            continue
        if rec[0] == "Change":
            mapping[rec[1]] = rec[2]
            continue
        mapping[rec[1]] = rec[2]
        tmp_answer.append([rec[1], rec[0]])
    for tmp in tmp_answer:
        if tmp[1] == "Enter":
            answer.append(f'{mapping[tmp[0]]}님이 들어왔습니다.')
        else:
            answer.append(f'{mapping[tmp[0]]}님이 나갔습니다.')
    return answer

File: js/오픈채팅방.js
function solution(records) {
  const idMap = new Map();
  const result = [];
  records.forEach(record => {
    const [action, id, nickname] = record.split(' ');
    if (action === 'Enter') {
      result.push({ id, action });
      idMap.set(id, nickname);
      return
    }

    if (action === 'Leave') {
      result.push({ id, action });
      return
    }
    
    if (action === 'Change') {
      idMap.set(id, nickname)
    }
  })
  return result.map(({id, action}) => {
    if (action === 'Enter') {
      return `${idMap.get(id)}님이 들어왔습니다.`;
    } else {
      return `${idMap.get(id)}님이 나갔습니다.`;
    }
  })
}